File: MonteCarloBSModel.fs
open System

let riskFreeRate = 0.01
let volatility = 0.2
let initialStockPrice = 100.0
let strikePrice = 100.0
let timeToMaturity = 1.0 // 1 year
let numberOfSimulations = 1000000

let generateGaussianNoise () =
    let u1 = Random().NextDouble()
    let u2 = Random().NextDouble()
    Math.Sqrt(-2.0 * Math.Log(u1)) * Math.Cos(2.0 * Math.PI * u2)

let simulateBlackScholes () =
    let drift = (riskFreeRate - 0.5 * volatility * volatility) * timeToMaturity
    let diffusion = volatility * Math.Sqrt(timeToMaturity)
    let mutable sumPayoff = 0.0

    for i in 1 .. numberOfSimulations do
        let gauss = generateGaussianNoise()
        let stockPriceAtMaturity = initialStockPrice * Math.Exp(drift + diffusion * gauss)
        let payoff = Math.Max(stockPriceAtMaturity - strikePrice, 0.0)
        sumPayoff <- sumPayoff + payoff

    Math.Exp(-riskFreeRate * timeToMaturity) * (sumPayoff / float numberOfSimulations)

[<EntryPoint>]
let main argv =
    let stopwatch = System.Diagnostics.Stopwatch.StartNew()
    
    let callPrice = simulateBlackScholes ()
    printfn "European Call Option Price = %f" callPrice

    stopwatch.Stop()
    printfn "Execution time: %f seconds" (float stopwatch.ElapsedMilliseconds / 1000.0)
    0

File: MonteCarloBSModel.cpp
#include <iostream>
#include <cmath>
#include <cstdlib>
#include <chrono>

double riskFreeRate = 0.01;
double volatility = 0.2;
double initialStockPrice = 100.0;
double strikePrice = 100.0;
double timeToMaturity = 1.0; // 1 year
long numberOfSimulations = 1000000;

double generateGaussianNoise() {
    double u = (double)rand() / RAND_MAX;
    double v = (double)rand() / RAND_MAX;
    return sqrt(-2.0 * log(u)) * cos(2.0 * M_PI * v);
}

double simulateBlackScholes() {
    double drift = (riskFreeRate - 0.5 * volatility * volatility) * timeToMaturity;
    double diffusion = volatility * sqrt(timeToMaturity);
    double sumPayoff = 0.0;

    for (long i = 0; i < numberOfSimulations; ++i) {
        double gauss = generateGaussianNoise();
        double stockPriceAtMaturity = initialStockPrice * exp(drift + diffusion * gauss);
        double payoff = std::max(stockPriceAtMaturity - strikePrice, 0.0);
        sumPayoff += payoff;
    }

    return exp(-riskFreeRate * timeToMaturity) * (sumPayoff / numberOfSimulations);
}

int main() {
    srand((unsigned)time(0));
    auto start = std::chrono::high_resolution_clock::now();

    double callPrice = simulateBlackScholes();
    std::cout << "European Call Option Price = " << callPrice << std::endl;

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;
    std::cout << "Execution time: " << elapsed.count() << " seconds\n";

    return 0;
}


